DirectProgramming/DPC++FPGA/Tutorials/Features/ac_int/README.md

```diff
@@ -0,0 +1,247 @@
+
+`ac_int` automatically promotes the result of all operations to the number of bits needed to represent all possible results without overflowing. For example, the addition of two 8-bit integers results in a 9-bit result to support overflow.
+
+However, if the user attempts to store the result in an 8-bit container, `ac_int` will let the user do this, but this leads to the discard of the extra carry bit. The responsibility lies on the user to use the correct datatype.
+
+These promotions rules are consistent across all architectures so the behavior should be equivalent on x86 or on FPGA.
+
+#### Shift Operation
+
+The behavior of a shift operation with an `ac_int` is slightly different from its behavior with native integer types. For full details, see the `ac_int` documentation in the file `ac_data_types_ref.pdf`. Some key points to remember are as follows:
+  - If the datatype of the shift amount is not explicitly `unsigned` (either using `ac_int<N, *false*>` or using the `unsigned` keyword), then the compiler will generate a more complex shifter that allows negative shifts and positive shifts. A right-shift by a negative amount is equivalent to a positive left-shift.
+  - Normally, you will not want to enable negative shifting, so you should use an `unsigned` datatype for the shift value to obtain a more resource efficient design.
+  - Shift values greater than the width of the data types are treated as a shift equal to the width of the datatype.
+  - The shift operation can be done more efficiently by specifying the amount to shift with the smallest possible `ac_int`.
+
+For example, in the tutorial, two kernels perform the left shift operation: `ShiftLeft` and `EfficientShiftLeft`. Both operate on an 14 bits wide `ac_int`. The former stores the shift amount in an `ac_int` which is 14 bits wide and the latter stores it in an `ac_int` which is 4 bits wide. The latter will generate simpler hardware.
+
+#### Bit Slice Operations
+
+The kernels `GetBitSlice` and `SetBitSlice` show how to read from and write to specific bits of an `ac_int` number. Note that only static bit widths are supported with such "slice" operations.
+
+For detailed documentation on the `set_slc` and `slc` APIs please see the file `ac_data_types_ref.pdf`
+
+## Key Concepts
+* The `ac_int` data-type can be used to generate hardware for only as many bits as is needed by the operation as compared to native integer types which generate hardware for the entire type width.
+* The left shift operation on `ac_int` can be implemented more efficiently when the amount to shift with is stored in a minimally sized ac_int``.
+* The `ac_int` data-type offers functions for several useful operations including reading and writing of certain bits of an `ac_int` number. This can be very useful in creating bit masks.
```

Comment:
I don't think 'creating bit masks' is what you mean to say here, but I'm not sure what you are trying to say . . .

Reply:
if it's so useful for creating masks, please show an example of creating a mask :)